3-enrich/beam-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich.beam/Enrich.scala

```diff
@@ -0,0 +1,205 @@
+      registry <- EnrichmentRegistry.parse(registryJson, false)(resolver).leftMap(_.toList.mkString("\n"))
+      _ <- if (emitPii(registry) && config.pii.isEmpty) {
+        "A pii topic needs to be used in order to use the pii enrichment".failure
+      } else {
+        ().success
+      }
+    } yield ParsedEnrichConfig(
+      config.input, config.output, config.bad, config.pii, resolverJson, registryJson)
+
+    parsedConfig match {
+      case Failure(e) =>
+        System.err.println(e)
+        System.exit(1)
+      case Success(config) =>
+        run(sc, config)
+        sc.close()
+    }
+  }
+
+  def run(sc: ScioContext, config: ParsedEnrichConfig): Unit = {
+    val cachedFiles: DistCache[List[Either[String, String]]] =
+      buildDistCache(sc, config.resolver, config.enrichmentRegistry)
+
+    val input: SCollection[Array[Byte]] = sc.pubsubSubscription(config.input).withName("input")
+    val enriched: SCollection[Validation[BadRow, EnrichedEvent]] = input
+      .map { rawEvent =>
+        cachedFiles()
+        implicit val resolver = ResolverSingleton.get(config.resolver)
+        val (enriched, time) = timeMs {
+          enrich(rawEvent, EnrichmentRegistrySingleton.get(config.enrichmentRegistry))
+        }
+        timeToEnrichDistribution.update(time)
+        enriched
+      }.withName("enriched")
+      .flatten.withName("enriched-flattened")
+
+    val (successes, failures) = enriched.partition(_.isSuccess)
+    val (tooBigSuccesses, properlySizedSuccesses) = successes
+      .collect { case Success(enrichedEvent) =>
+        getEnrichedEventMetrics(enrichedEvent)
+          .foreach(ScioMetrics.counter(MetricsNamespace, _).inc())
+        val formattedEnrichedEvent = tabSeparatedEnrichedEvent(enrichedEvent)
+        val size = getStringSize(formattedEnrichedEvent)
+        enrichedEventSizeDistribution.update(size.toLong)
+        (formattedEnrichedEvent, size)
+      }.withName("enriched-successes")
+      .partition(_._2 >= MaxRecordSize)
+    properlySizedSuccesses.withName("undersized-enriched-successes")
+      .map(_._1).withName("enriched-good")
+      .saveAsPubsub(config.output)
+
+    val resolver = ResolverSingleton.get(config.resolver)
+    val registry = EnrichmentRegistrySingleton.get(config.enrichmentRegistry)(resolver)
+    val piis = if (emitPii(registry)) {
+      val (tooBigPiis, properlySizedPiis) = successes
+        .collect { case Success(enrichedEvent) =>
+          getPiiEvent(enrichedEvent)
+            .map(tabSeparatedEnrichedEvent)
+            .map(formatted => (formatted, getStringSize(formatted)))
+        }
+        .flatten.withName("pii-successes")
+        .partition(_._2 >= MaxRecordSize)
+      Some((tooBigPiis, properlySizedPiis))
+    } else {
+      None
+    }
+
+    (piis |@| config.pii) { (piis, pii) =>
+      val properlySizedPiis = piis._2
+      properlySizedPiis.withName("undersized-pii-successes")
```

Comment:
Undersized sounds like there is something wrong with them (unless I misunderstood the code and there is something wrong)